onnxruntime/core/providers/cuda/tensor/transpose.cc

```diff
@@ -92,20 +92,76 @@ Status Transpose::DoTranspose(const Transpose& kernel,
   const std::vector<int64_t>& output_dims = output.Shape().GetDims();
 
   auto rank = static_cast<int32_t>(input_dims.size());
-  TensorPitches original_input_strides(input_dims);
-  TensorPitches original_output_strides(output_dims);
 
-  TArray<int64_t> input_strides(rank);
-  for (auto i = 0; i < rank; i++) {
-    input_strides[i] = original_input_strides[permutations[i]];
-  }
-  TArray<fast_divmod> output_strides(rank);
-  for (auto i = 0; i < rank; i++) {
-    output_strides[i] = fast_divmod(gsl::narrow_cast<int>(original_output_strides[i]));
+  // flatten the adjacent dimensions which are contiguous
+  // for example: permutations[0, 2, 3, 1] -> [0, 2, 1], permutations[0, 3, 1, 2] -> [0, 2, 1]
```

Comment:
>permutations[0, 3, 1, 2] -> [0, 2, 1] [](start = 57, length = 37)

i don't understand this example:
```
permutations[0, 3, 1, 2] -> [0, 2, 1]
```
why is it not [0, 3, 1]?

Reply:
For this example, 1 and 2 can be flattened and become [0, 3, 1], then index 3 needs to be decreased by 1 since the number of dimensions has been reduced from 4 to 3, then the new permutations is [0, 2, 1].